include/pybind11/attr.h

```diff
@@ -148,14 +150,49 @@ struct type_record {
     /// Function pointer to class_<..>::dealloc
     void (*dealloc)(PyObject *) = nullptr;
 
-    // Pointer to RTTI type_info data structure of base class
-    const std::type_info *base_type = nullptr;
-
-    /// OR: Python handle to base class
-    handle base_handle;
+    /// List of base classes of the newly created type
+    list bases;
 
     /// Optional docstring
     const char *doc = nullptr;
+
+    PYBIND11_NOINLINE void add_base(const std::type_info *base, void *(*caster)(void *)) {
+        auto base_info = detail::get_type_info(*base, false);
+        if (!base_info) {
+            std::string tname(base->name());
+            detail::clean_type_id(tname);
+            pybind11_fail("generic_type: type \"" + std::string(name) +
+                          "\" referenced unknown base type \"" + tname + "\"");
+        }
+
+        bases.append((PyObject *) base_info->type);
+
+        if (caster)
+            base_info->implicit_casts.push_back(std::make_pair(type, caster));
+    }
+};
+
+/// Helper class used to populate type_record::bases from a list of template arguments
+template <template<typename> class Predicate, typename Child, typename... Args> struct add_base_helper { };
+template <template<typename> class Predicate, typename Child, typename Base, typename... Args>
+struct add_base_helper<Predicate, Child, Base, Args...> {
+    template <typename T = Base, std::enable_if_t<Predicate<T>::value, int> = 0>
+    static void apply(type_record &rec) {
+        rec.add_base(&typeid(Base), [](void *src) -> void * {
+            return static_cast<Base *>(reinterpret_cast<Child *>(src));
+        });
+        add_base_helper<Predicate, Child, Args...>::apply(rec);
+    }
+
+    template <typename T = Base, std::enable_if_t<!Predicate<T>::value, int> = 0>
+    static void apply(type_record &rec) {
+        add_base_helper<Predicate, Child, Args...>::apply(rec);
+    }
+};
+
+template <template<typename> class Predicate, typename Child>
+struct add_base_helper<Predicate, Child> {
+    static void apply(detail::type_record &) { }
```

Comment:
I haven't actually compiled the following snippet, so take it with a grain of salt, but it's a possible simplification (non-recursive, non-SFINAE):
```c++
template <typename Base, typename Child>
void add_base_helper(type_record &rec, std::true_type /*is_base*/) {
    rec.add_base(&typeid(Base), [](void *src) -> void * {
        return static_cast<Base *>(reinterpret_cast<Child *>(src));
    });
}

template <typename...> void add_base_helper(type_record &, std::false_type /*is_base*/) { }
```
Called as:
```c++
add_base_helper<options, type_>(record, is_base_class<options>())...
```
(Plus the usual variadic call `{(f(), 0)...}` dance.)